### PR TITLE
Making connections even more dynamic and ready for new chains.

### PR DESCRIPTION
**Pull request by hbulgarini**
Address some of the comments of PR #136 
But also it address the first version of what it was proposed on #139

### CODE REVIEW

File: src/util/tests/envVariablesValidations.test.ts [hunk 1]
```diff
@@ -75,8 +75,4 @@ describe('Validate .env file', () => {
   test('expected ENV variables are set', () => {
     expect(checkExpectedVariables()).toBeTruthy();
   });
-
```
Comment (hbulgarini) [inline src/util/tests/envVariablesValidations.test.ts:78]: Needed to remove this since now some variables are optional.

File: src/hooks/useApiChainName.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+//
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges UI.  If not, see <http://www.gnu.org/licenses/>.
+
+import { ApiPromise } from '@polkadot/api';
+import { useEffect, useState } from 'react';
+import logger from '../util/logger';
+import { getChainName } from '../util/getConfigs';
+import { ConnectionChainInformation } from '../types/sourceTargetTypes';
+import { BASE_CUSTOM_TYPES } from '../constants';
```
Comment (tomusdrw) [inline src/hooks/useApiChainName.ts:22]: I'm a bit confused, why do we still need `BASE_CUSTOM_TYPES` and `useApiChainName` hook at all? Since that with your decision to put custom types URLs to ENV variables (which is not something that I suggested in other PR, but I do like (cause it's even more configurable)) we can create API connection right away with proper types?

File: src/hooks/useApiChainName.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      await apiPromise
+        .disconnect()
+        .catch((err) => logger.error(`Chain ${chainNumber}: Error disconnecting. Details: ${err.message}`));
+      logger.info(`Chain ${chainNumber} disconnected`);
+    };
+
+    if (apiPromise.isConnected && chainName) {
+      disconnect();
+    }
+  }, [apiPromise, chainName, chainNumber]);
```
Comment (tomusdrw) [inline src/hooks/useApiChainName.ts:59]: I'm really struggling to understand why we need this file at all, but even if we do, splitting the connect-disconnect cycle is imho a bad idea Why not simply have that as one single `useEffect`, where you basically:
1. Connect
2. Set State (watch out for mounted state,  [look for `let ignore = ` in React docs](https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies))
3. Disconnect.

File: src/hooks/useConnections.ts [hunk 4]
```diff
@@ -22,14 +22,23 @@ import { SourceTargetState, ChainDetails, ConnectionChainInformation, Configs }
 interface Props {
   connectionDetailsOne: ConnectionChainInformation;
   connectionDetailsTwo: ConnectionChainInformation;
+  initConnectionDetailsOne: ConnectionChainInformation;
+  initConnectionDetailsTwo: ConnectionChainInformation;
```
Comment (tomusdrw) [inline src/hooks/useConnections.ts:26]: What's wrong with `1` and `2` that we have to actually spell them? Also I'd very much prefer if we used `arrays` everywhere, at some point we might need to connect to `three` or even `dozen` of other chains and adding more configuration variables to this interface is not the right way to go.

File: src/types/sourceTargetTypes.ts [hunk 5]
```diff
@@ -26,7 +26,7 @@ export interface ApiPromiseConnectionType {
 
 export interface ConnectionChainInformation {
   chainNumber: string;
-  hasher?: (data: Uint8Array) => Uint8Array;
+  hasher: ((data: Uint8Array) => Uint8Array) | undefined;
```
Comment (tomusdrw) [inline src/types/sourceTargetTypes.ts:29]: Is this required? It's almost exactly the same, but just less convenient (you always require the `hasher` field, but in case it's not needed it must have `undefined` value).

File: src/util/getConfigs.ts [hunk 6]
```diff
@@ -20,10 +20,15 @@ export const getConfigs = async (apiPromise: ApiPromise) => {
...
-  const chainName = systemChain.split(' ')[0];
   const bridgeIds = prop.get('bridgeIds');
 
-  return { bridgeId: [], bridgeIds, chainName, ss58Format: parseInt(ss58Format.toString()) };
+  return { bridgeId: [], bridgeIds, ss58Format: parseInt(ss58Format.toString()) };
+};
+
+export const getChainName = async (apiPromise: ApiPromise) => {
+  const systemChain = await apiPromise.rpc.system.name();
+  const chainName = systemChain.split(' ')[0];
```
Comment (tomusdrw) [inline src/util/getConfigs.ts:31]: What happens here? We should just take the chain name as-is.

File: src/util/tests/encodeUrl.test.ts [hunk 7]
```diff
@@ -18,7 +18,7 @@ import { stringToU8a } from '@polkadot/util';
 import { base64Encode } from '@polkadot/util-crypto';
 import { zlibSync } from 'fflate';
 
-import customTypesRialto from '../../configs/customTypesRialto.json';
+import customTypesRialto from '../../configs/substrateCustomTypes/customTypesChain1.json';
```
Comment (tomusdrw) [inline src/util/tests/encodeUrl.test.ts:21]: This might break the test in case different env variables are set. I'd suggest to actually fetching the rialto types to make sure it's deterministic.

File: src/util/tests/envVariablesValidations.test.ts (hunk 1 above)
Comment (tomusdrw) [reply src/util/tests/envVariablesValidations.test.ts:78]: We can instead list the optional ones?

File: src/util/tests/envVariablesValidations.test.ts [hunk 8]
```diff
@@ -17,7 +17,7 @@
 import dotenv from 'dotenv';
 
 import { checkEnvVariable } from '../envVariablesValidations';
-import { checkExpectedVariables, checkUnexpectedVariables } from '../envVariablesValidations';
```
Comment (tomusdrw) [inline src/util/tests/envVariablesValidations.test.ts:20]: If you remove the function import and it's literally the only place where that function is used, please remove the function as well (but note I don't think it should be removed, just altered to cater for optional vars.)

**Review by tomusdrw — COMMENTED** [review]
It's a tiny bit different than what we discussed initially, but it's actually better than the original proposal (which was suggesting to hardcode the list of "known chains").

However there are couple of code improvement suggestions.

**Review by hbulgarini — COMMENTED** [review]
Comments addressed.

File: src/util/envVariablesValidations.ts [hunk 9]
```diff
@@ -15,13 +15,19 @@
...
   'REACT_APP_SUBSTRATE_PROVIDER_CHAIN_2',
   'REACT_APP_LANE_ID',
   'REACT_APP_KEYRING_DEV_LOAD_ACCOUNTS'
 ];
 
+const optionalVariables = [
+  'REACT_APP_CUSTOM_TYPES_URL_CHAIN_1',
+  'REACT_APP_CUSTOM_TYPES_URL_CHAIN_2',
+  'REACT_APP_CUSTOM_HASHER_CHAIN_1',
+  'REACT_APP_CUSTOM_HASHER_CHAIN_2'
```
Comment (tomusdrw) [inline src/util/envVariablesValidations.ts:28]: Cool stuff. Wdyt about chaing the variable names though to have `CHAIN_1` in front, so that you can easily sort them:
```
REACT_APP_CHAIN_1_CUSTOM_TYPES_URL_=https://raw.githubusercontent.com/paritytech/parity-bridges-common/master/deployments/types-rialto.json
REACT_APP_CHAIN_1_SUBSTRATE_PROVIDER=wss://wss.rialto.brucke.link
REACT_APP_CHAIN_2_CUSTOM_TYPES_URL=https://raw.githubusercontent.com/paritytech/parity-bridges-common/master/deployments/types-millau.json
REACT_APP_CHAIN_2_SUBSTRATE_PROVIDER=wss://wss.millau.brucke.link
REACT_APP_CHAIN_2_CUSTOM_HASHER_=blake2Keccak256Hasher

REACT_APP_LANE_ID=0x00000000
REACT_APP_KEYRING_DEV_LOAD_ACCOUNTS=false
```
The reason for that is that I think it's easier to think "per chain" instead of "per functionality". Like when you start defining your env variables, you will most likely go with:
1. What are the things for one chain I need to configure?
2. What are the thigns for second chain I need to configure?

Instead of:
1. What are the custom types for chain1 and chain2?
2. What are the custom hashers for chain1 and chain2?
3. What are the URLs for chain1 and chain2?

Basically - it's easier to enumerate chains than it is to enumerate all possible configuration options from a developer perspective that is configuring the UI.

Comment (hbulgarini) [reply src/util/envVariablesValidations.ts:28]: Yeap. Sounds good to me

File: .env [hunk 10]
```diff
@@ -1,5 +1,11 @@
...
+REACT_APP_CHAIN_1_SUBSTRATE_PROVIDER=wss://wss.rialto.brucke.link
+REACT_APP_CHAIN_2_CUSTOM_HASHER=blake2Keccak256Hasher
+REACT_APP_CHAIN_2_CUSTOM_TYPES_URL=https://raw.githubusercontent.com/paritytech/parity-bridges-common/master/deployments/types-millau.json
+REACT_APP_CHAIN_2_SUBSTRATE_PROVIDER=wss://wss.millau.brucke.link
+
 REACT_APP_LANE_ID=0x00000000
 REACT_APP_KEYRING_DEV_LOAD_ACCOUNTS=false
+
+
+
```
Comment (tomusdrw) [inline .env:11]:
```suggestion
```

**Review by tomusdrw — APPROVED** [review]
The hasher type would be good to solve, but the code looks really good already, thanks for bearing with me :)

**Comment by hbulgarini** [comment]
> The hasher type would be good to solve, but the code looks really good already, thanks for bearing with me :)

Solved. Thanks for your patience and help.
